### PR TITLE
Support for ~/.chplconfig file

### PR DESCRIPTION
**Pull request by ben-albrecht**
# chplconfig

This PR adds support for a `chplconfig`/`.chplconfig` file, which allows users to override the default environment variables set by `printchplenv`.

## Syntax

The file syntax is simple. Users can define variables with the following format:

```
CHPL_ENV = VALUE
```

Above, `CHPL_ENV` will be set to `VALUE` with any additional white space stripped away.

Any blank line, or line with a first character of `#` will be considered a comment.

An example `~/.chplconfig` , with sample comments:

```
# ~/.chplconfig

# Default to multi-locale
CHPL_COMM = gasnet

# Always use qthreads
CHPL_TASKS = qthreads

# System GMP is available on these machines
CHPL_GMP = system
```

## Generating a `chplconfig`

The syntax has been chosen in such a way that a user can dump their current configuration into a `chplconfig` file with `printchplenv --simple`, e.g.

```bash
printchplenv --simple > ~/.chplconfig
```

A new flag, `printchplenv --overrides`, has also been added. This flag will print the environment variables that have been overridden by either an environment variables or `chplconfig` file.

A user can dump their current overrides into a `chplconfig` file as well:

```bash
printchplenv --overrides > ~/.chplconfig
```

## Search Paths

The search priority for `chplconfig` files is as follows:

1. `$CHPL_CONFIG`
2. `$HOME` (`~/`)
3. `$CHPL_HOME`

When both a `chplconfig` and `.chplconfig` is defined, the visible `chplconfig` will be prioritized.

Only a single `chplconfig` file will be used, i.e. as soon as we find a valid `chplconfig` we stop searching and use that.

Note that `$CHPL_CONFIG` variable is the path to the *enclosing* directory - not the full path including `chplconfig` itself.

## Variable priority

Variable precedence goes in the following order:

1. compiler flags: `chpl --comm=gasnet`
2. environment variables: `CHPL_COMM=gasnet`
3. `chplconfig` file: *example shown above*
4. `printchplenv` defaults: `printchplenv --simple`

## Warnings

Warnings will be displayed under the following conditions:

* A user defines a `$CHPL_CONFIG`, but no `.chplconfig` or `chplconfig` file is found within it's path

* A `chplconfig` line is not a comment and does not have a single `=` character

* A user defines a `$CHPL_*` variable that is not one of the (now) 23 environment variables printed in `printchplenv --simple`

* A user defines a variable multiple times

## Testing

Added 3 new tests (1 for correctness, 2 for warnings), and consolidated  `chplenv` related tests into a `test/chplenv/` directory.

*paratest passes*

## Future Features

* Documentation
    * man page
    * online docs page

* Modify build system to dump a `builtconfig` file in the `lib/*` directories, providing the configuration in which a given runtime was built.

### CODE REVIEW

File: util/chplenv/chpl_arch.py [hunk 1]
```diff
@@ -8,7 +8,7 @@
 chplenv_dir = os.path.dirname(__file__)
 sys.path.insert(0, os.path.abspath(chplenv_dir))
 
-import chpl_comm, chpl_compiler, chpl_platform, utils
+import chpl_comm, chpl_compiler, chpl_platform, utils, defaults
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: Should `defaults` be alphabetized in this list?

**Comment by ronawho** [comment]
Should update the [template] (https://github.com/chapel-lang/chapel/blob/master/util/chplenv/template) as well

File: util/chplenv/utils.py [hunk 2]
```diff
@@ -19,15 +25,15 @@ def memoize_wrapper(*args, **kwargs):
...
         dirname = os.path.dirname
         chpl_home = dirname(dirname(dirname(os.path.realpath(__file__))))
     return chpl_home
 
 @memoize
 def using_chapel_module():
-    chpl_home = os.environ.get('CHPL_HOME', '')
+    chpl_home = defaults.get('CHPL_HOME', '')
     if chpl_home != '':
         return chpl_home == os.environ.get('CHPL_MODULE_HOME', '')
```
Comment (ronawho) [inline util/chplenv/utils.py]: I *think* you should be using defaults.get here as well, but I'm not convinced if that's true or not

File: test/chplenv/chplconfig/correctness/correctness.good [hunk 3]
```diff
@@ -0,0 +1,17 @@
+CHPL_TARGET_PLATFORM: darwin *
```
Comment (ronawho) [inline test/chplenv/chplconfig/correctness/correctness.good]: I'm not following how these tests work, will need to bug you in person about them

File: util/chplenv/defaults.py [hunk 4]
```diff
@@ -0,0 +1,155 @@
+#!/usr/bin/env python
+"""
+Check .chplconfig file and environment variables for default value overrides
+"""
+from __future__ import print_function
+import os
+import sys
+from sys import stderr
```
Comment (ronawho) [inline util/chplenv/defaults.py]: You don't actually use stderr very much, so I have a slight preference of just calling `sys.stderr` and removing this import. I can be talked out of that if you care though

File: util/chplenv/utils.py (hunk 2 above)
Comment (ben-albrecht) [inline util/chplenv/utils.py]: I purposefully did not include CHPL_MODULE_HOME since it is not one of the 23 env vars printed with `printchplenv --simple`. We could add it to the list though.

File: util/chplenv/defaults.py [hunk 5]
```diff
@@ -0,0 +1,155 @@
...
+sys.path.insert(0, os.path.abspath(chplenv_dir))
+
+# Global dictionary that will be populated w/ .chplconfig file contents
+chplconfig = {}
+
+# Globally track if we've parsed already
+parsed = False
+
+# Valid options for override by .chplconfig file
+CHPLS = [
```
Comment (ronawho) [inline util/chplenv/defaults.py]: Why do you need this array? Can't you just stuff whatever variables are in the .chpconfig file into the dictionary? My concern is that it'll be easy to miss updating this list if a new var is added.

File: util/chplenv/defaults.py [hunk 6]
```diff
@@ -0,0 +1,155 @@
...
+
+
+def parse():
+    """ Parse .chplconfig file for acceptable env var overrides """
+
+    # Default initialized to empty dictionary
+    chplconfig = {}
+
+    # Potential paths of .chplconfig file
+    chplconfigpath = os.environ.get('CHPL_CONFIG')
```
Comment (ronawho) [inline util/chplenv/defaults.py]: I'd prefer to put the "Find .chplconfig file" in a separate function

File: test/chplenv/chplconfig/correctness/correctness.good (hunk 3 above)
Comment (ben-albrecht) [inline test/chplenv/chplconfig/correctness/correctness.good]: The `.chpl` file is the `.good`, which prints out the expected output of the test. The `.precomp` is the actual test, which calls `printchplenv` after setting `$CHPL_CONFIG` to the path of the current directory.

This particular test confirms that `printchplenv` recognized that all defaults were overridden and places a `*` next to their name. I had to use actual values, e.g. `CHPL_TARGET_PLATFORM=darwin`, for the overrides so that I didn't get any other random `printchplenv` errors, such as `"Unrecognized CHPL_TARGET_PLATFORM"`.

File: util/chplenv/chpl_arch.py (hunk 1 above)
Comment (ben-albrecht) [inline util/chplenv/chpl_arch.py]: Should have seen that one coming...

File: util/chplenv/defaults.py (hunk 6 above)
Comment (ben-albrecht) [inline util/chplenv/defaults.py]: Added a `_find()` and `_prettypath()` (since that was defined when searching for path)

File: util/chplenv/defaults.py (hunk 4 above)
Comment (ben-albrecht) [inline util/chplenv/defaults.py]: noted

File: util/chplenv/defaults.py (hunk 5 above)
Comment (ben-albrecht) [inline util/chplenv/defaults.py]: right - moved into `utils.py`

File: util/chplenv/defaults.py [hunk 7]
```diff
@@ -0,0 +1,170 @@
+#!/usr/bin/env python
+"""
+Provides simple 'get()' interface for accessing default value overrides
+Checks environment variables first, then ..chplconfig file for definitions
```
Comment (ronawho) [inline util/chplenv/defaults.py]: `..chplconfig` => `.chplconfig`?

File: util/chplenv/defaults.py [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+        self.chplconfig = {}
+
+        # List of warnings to track if a warning has occurred
+        self.warnings = []
+
+        # Find chplconfigfile and prettypath
+        self.find()
+
+        # Populate chplconfig
+        if self.chplconfigfile:
```
Comment (ronawho) [inline util/chplenv/defaults.py]: Should you set `self.chplconfigfile = None` above?

File: util/chplenv/defaults.py [hunk 9]
```diff
@@ -0,0 +1,170 @@
...
+            return True
+
+        # Warn about duplicate entries, but don't skip, just overwrite
+        elif var in self.chplconfig.keys():
+            self.warnings.append(
+            (
+                'Warning: {0}:line {1}: '
+                'Duplicate entry of "{2}"\n'
+            ).format(self.prettypath, linenum, var))
+            return False
```
Comment (ronawho) [inline util/chplenv/defaults.py]: I'd remove this `return False`

File: util/chplenv/defaults.py [hunk 10]
```diff
@@ -0,0 +1,180 @@
...
+        """ Wrapper for chplconfig[var] intended to mirror os.environ.get() """
+        if var in self.chplconfig.keys():
+            return self.chplconfig[var]
+
+    def find(self):
+        """ Find chplconfig file path"""
+
+        # Places to look for a chplconfig file
+        chplconfigpath = os.environ.get('CHPL_CONFIG')
+        home = os.path.expanduser('~')
```
Comment (ronawho) [inline util/chplenv/defaults.py]: `home` is only used once, maybe just say `ccpath = os.path.expanduser('~')` below instead?

File: util/chplenv/defaults.py [hunk 11]
```diff
@@ -0,0 +1,180 @@
...
+                    continue
+
+                var, val = [f.strip() for f in fields]
+
+                if self.invalid_entry(var, val, linenum):
+                    continue
+
+                self.chplconfig[var] = val
+
+    def invalid_assignment(self, fields, linenum):
```
Comment (ronawho) [inline util/chplenv/defaults.py]: `invalid_assignment` is a slightly odd name to me. I'd prefer `is_invalid_assignment` same with `invalid_entry`

File: util/chplenv/defaults.py [hunk 12]
```diff
@@ -0,0 +1,180 @@
...
+        """ Check if formatting of fields is formatted like 'ENV = VAR' """
+
+        # Check if line is a comment (has no '=')
+        if len(fields) < 2:
+            return True
```
Comment (ronawho) [inline util/chplenv/defaults.py]: It's a little odd that a comment is labeled as an "invalid_assignment"

File: util/chplenv/defaults.py (hunk 9 above)
Comment (ben-albrecht) [inline util/chplenv/defaults.py]: Fair enough, I left it in there to keep the format consistent, but it's admittedly redundant and not worth the effort.

File: util/chplenv/defaults.py (hunk 10 above)
Comment (ben-albrecht) [inline util/chplenv/defaults.py]: Keeping this in for clarify/organization

**Comment by gbtitus** [comment]
I would call it .chplrc instead of .chplconfig.  The reason for the 'rc' ending is probably not widely known anymore (and I'm too lazy right now to even look it up), still, folks are used to it I think.

Could we search $CHPL_HOME for the config file instead of $CHPL_CONFIG?  It seems like an obvious use case is to associate different default settings with different repo clones.  As an alternative, could the search procedure first look in $CHPL_HOME, then $CHPL_CONFIG, then $HOME?

The syntax has been simplified to the point of being dangerous.  :-)   As it stands, what to a human is obviously a comment
    `;; Here we're assuming CHPL_COMM=gasnet`
will actually be parsed (and then griped about) because the '=' makes it looks like an assignment, to the config file parser.  It would be better to say that blank lines or lines starting with optional white space and then '#' are comments, and everything else has to match assignment syntax.  This would open up the possibility of using the same file as a Chapel config file or a shell script, what a bonus!

**Comment by ben-albrecht** [comment]
Hi @gbtitus -- thanks for the feedback.

> I would call it .chplrc instead of .chplconfig. The reason for the 'rc' ending is probably not widely known anymore (and I'm too lazy right now to even look it up), still, folks are used to it I think.

We chose `chplconfig` because (1) rc is short for *run commands*, which doesn't completely make sense in this context, and (2) it appears that other language-level configuration files use a `conf`/`config` style over an `rc` style. I agree that `*rc` is the norm for broader applications though.

>As an alternative, could the search procedure first look in $CHPL_HOME, then $CHPL_CONFIG, then $HOME?

Yeah, I like this.

>  It would be better to say that blank lines or lines starting with optional white space and then '#' are comments, and everything else has to match assignment syntax. 

I can do that. I chose the current implementation out of simplicity, both in implementation and interface, but I agree that users could get confused by it.

**Comment by gbtitus** [comment]
> We chose chplconfig because (1) rc is short for run commands, which
> doesn't completely make sense in this context, and (2) it appears that
> other language-level configuration files use a conf/config style

Ah, thanks for the reminder.  .chplconfig or chplconfig it is, then.

**Comment by gbtitus** [comment]
I took a brief final glance and this looks good.

File: util/chplenv/chpl_arch.py [hunk 13]
```diff
@@ -224,10 +225,10 @@ def get_cpuinfo(platform='linux'):
     vendor_string = ''
     feature_string = ''
     if platform == "darwin":
-        vendor_string = utils.run_command(['sysctl',
+        vendor_string = run_command(['sysctl',
                                            '-n',
```
Comment (ronawho) [inline util/chplenv/chpl_arch.py]: Need to fix indentation of the lines not that `utils.` is gone

File: util/chplenv/compiler_utils.py [hunk 14]
```diff
@@ -0,0 +1,49 @@
...
+from utils import memoize
+
+
+@memoize
+def get_compiler_version(compiler):
+    version_string = '0'
+    if 'gnu' in compiler:
+        # Asssuming the 'compiler' version matches the gcc version
+        # e.g., `mpicc -dumpversion == gcc -dumpversion`
+        version_string = run_command(['gcc', '-dumpversion'])
```
Comment (ronawho) [inline util/chplenv/compiler_utils.py]: Need to import run_command

File: util/chplenv/__init__.py [hunk 15]
```diff
@@ -19,6 +19,10 @@
     'chpl_timers',
     'chpl_unwind',
     'chpl_wide_pointers',
+    # et cetera
+    'chpl_home_utils',
+    'compiler_utils',
+    'overrides',
```
Comment (ronawho) [inline util/chplenv/__init__.py]: alphabetize?

File: util/chplenv/template [hunk 16]
```diff
@@ -3,24 +3,28 @@
...
+
+import defaults, utils
 from utils import memoize
 
 #TODO: @memoize causes the calls to this function to get cached
 @memoize
 def get(flag='host'):
     if flag == 'host':
-        FOOBAR_val = os.environ.get('CHPL_HOST_FOOBAR', '')
+        FOOBAR_val = defaults.get('CHPL_HOST_FOOBAR', '')
```
Comment (ronawho) [inline util/chplenv/template]: Still using defaults instead of overrides

Comment (ronawho) [inline util/chplenv/template]: Maybe the printchplenv tests should check the template to make sure it works?

Or you'll probably argue we should just get rid of the template

File: util/chplenv/chpl_tasks.py [hunk 17]
```diff
@@ -20,7 +21,7 @@ def get():
         # use the qthreads it provides even if the user has an older CCE loaded
         using_qthreads_incompatible_cce = False
         if compiler_val == 'cray-prgenv-cray':
-            if (utils.get_compiler_version(compiler_val) < CompVersion('8.4') and
+            if (get_compiler_version(compiler_val) < CompVersion('8.4') and
                     not using_chapel_module()):
```
Comment (ronawho) [inline util/chplenv/chpl_tasks.py]: Need to import using_chapel_module from utils. huh, I guess that was broken before too

File: util/chplenv/compiler_utils.py [hunk 18]
```diff
@@ -0,0 +1,49 @@
+""" Compiler utility functions for chplenv modules """
```
Comment (ronawho) [inline util/chplenv/compiler_utils.py]: "Compiler" => "Backend compiler" since it's not chapel compiler utils?

File: util/chplenv/overrides.py [hunk 19]
```diff
@@ -0,0 +1,232 @@
...
+
+# Global instance that contains parsed chplconfig assignments
+chplconfig = ChapelConfig()
+
+
+@memoize
+def get(var, default=None):
+    """ Check if variable has a default defined somewhere """
+    # Check environment variable
+    value = os.environ.get(var)
```
Comment (ronawho) [inline util/chplenv/overrides.py]: I'd be tempted to suggest just writing this function as 

```py
return os.environ.get(var) or chplconfig.get(var) or default
```

File: util/chplenv/__init__.py (hunk 15 above)
Comment (ben-albrecht) [inline util/chplenv/__init__.py]: Will alphabetize within this section, but keeping these util modules separate from others, similar to the third party helpers.

File: util/chplenv/chpl_tasks.py (hunk 17 above)
Comment (ben-albrecht) [inline util/chplenv/chpl_tasks.py]: fixed

File: util/chplenv/compiler_utils.py (hunk 14 above)
Comment (ben-albrecht) [inline util/chplenv/compiler_utils.py]: fixed

File: util/chplenv/template (hunk 16 above)
Comment (ben-albrecht) [inline util/chplenv/template]: Your speculation is correct, but if we end up doing this, I think it belongs in a different PR.

**Comment by ronawho** [comment]
👍 👍
